include/deal.II/meshworker/mesh_loop.h

```diff
@@ -0,0 +1,313 @@
+   * AssembleFlags `flags` that are passed. The `cell_worker` is passed the
+   * cell identifier, a ScratchData object, and a CopyData object, following
+   * the same principles of WorkStream::run. Internally the function passes to
+   * `boundary_worker`, in addition to the above, also a `face_no` parameter
+   * that identifies the face on which the integration should be performed. The
+   * `face_worker` instead need to identify the current face unambiguously both on
+   * the cell and on the neighboring cell, and it is therefore called with six
+   * arguments (three for each cell: the actual cell, the face index, and
+   * the subface_index. If no subface integration is needed, then the
+   * subface_index is numbers::invalid_unsigned_int) in addition to the usual
+   * ScratchData and CopyData objects.
+   *
+   * If the flag AssembleFlags::assemble_own_cells is passed, then the default
+   * behaviour is to first loop over faces and do the work there, and then
+   * compute the actual work on the cell. It is possible to perform the
+   * integration on the cells before working on faces, by adding the flag
+   * AssembleFlags::cells_first.
+   *
+   * If the flag AssembleFlags::assemble_own_interior_faces_once is specified,
+   * then each interior face is visited only once, and the `face_worker` is
+   * assumed to integrate all face terms at once (and add contributions to both
+   * sides of the face in a discontinuous Galerkin setting).
+   *
+   * This method is equivalent to the WorkStream::run() method when
+   * AssembleFlags contains only `assemble_own_cells`, and can be used as a
+   * drop-in replacement for that method.
+   *
+   * The two data types ScratchData and CopyData need to have a working copy
+   * constructor. ScratchData is only used in the worker function, while CopyData is
+   * the object passed from the worker to the copier.
+   *
+   * The queue_length argument indicates the number of items that can be live at any
+   * given time. Each item consists of chunk_size elements of the input stream that
+   * will be worked on by the worker and copier functions one after the other on the
+   * same thread.
+   *
+   * If your data objects are large, or their constructors are expensive, it is
+   * helpful to keep in mind that queue_length copies of the ScratchData object
+   * and queue_length*chunk_size copies of the CopyData object are generated.
+   *
+   * @ingroup MeshWorker
+   * @author Luca Heltai and Timo Heister, 2017
+   */
+  template <class CellIteratorType,
+            class ScratchData, class CopyData>
+  void mesh_loop(const CellIteratorType &begin,
+                 const typename identity<CellIteratorType>::type &end,
+
+                 const typename identity<std::function<void (const CellIteratorType &, ScratchData &, CopyData &)>>::type &cell_worker,
+                 const typename identity<std::function<void (const CopyData &)>>::type &copier,
```

Comment:
Here and below, that's a funny way to place the `identity` wrapper. I suppose it's fine like this, but what you really meant to do was to apply it to the `CellIterator` and `CopyData` types. I agree that it's easier to read this way.